[WPE] Web Platform: implement WebPageProxy::callAfterNextPresentationUpdate

carlosgcampos opened the pull request:
#### 1512e48affb3304f86b9b4cbbcb2b934e594faf5
<pre>
[WPE] Web Platform: implement WebPageProxy::callAfterNextPresentationUpdate
<a href="https://bugs.webkit.org/show_bug.cgi?id=269353">https://bugs.webkit.org/show_bug.cgi?id=269353</a>

Reviewed by Adrian Perez de Castro.

With the old API we just use a 0 seconds timer, but with the new API we
can use the buffer-renderer signal of WPEView to actually call the
callback after the next buffer is rendered. This will hopefully reduce
the flakiness of ref tests.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):

Canonical link: <a href="https://commits.webkit.org/274710@main">https://commits.webkit.org/274710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89605c7481fec6331a54506180566ad1540ad8d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42383 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16180 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40413 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/15885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13748 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43662 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36175 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35733 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39509 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12051 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16273 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16321 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5241 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -958,4 +959,25 @@ void View::setCursor(const WebCore::Cursor& cursor)
+            auto& webView = *reinterpret_cast<View*>(userData);
+            g_signal_handler_disconnect(view, std::exchange(webView.m_bufferRenderedID, 0));
+            webView.m_nextPresentationUpdateCallback();
+        }), this);
```
AFAIU while GLib signals should be fast, setup/teardown can be slow due to looking signal information by their name, setting up marshalling, closures, etc. and there are locks involved.

Maybe we should connect this once on first use and then only check whether `webView.m_nextPresentationUpdateCallback` is non-null before use in the signal handler callback.

mcatanzaro replied:
I think signals are not fast? I would generally not emit a GObject signal in performance-sensitive code (unless you're confident the impact is not too bad).

mcatanzaro replied:
(The usual workaround, if it turns out to be too slow, would be to register a callback instead.)

aperezdc reviewed (approved):
The logic LGTM, but I left a comment about signal setupo/teardown.

carlosgcampos committed:
[WPE] Web Platform: implement WebPageProxy::callAfterNextPresentationUpdate
https://bugs.webkit.org/show_bug.cgi?id=269353

Reviewed by Adrian Perez de Castro.

With the old API we just use a 0 seconds timer, but with the new API we
can use the buffer-renderer signal of WPEView to actually call the
callback after the next buffer is rendered. This will hopefully reduce
the flakiness of ref tests.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):

Canonical link: https://commits.webkit.org/274710@main
